opentelemetry-sdk/src/metrics/periodic_reader.rs

```diff
@@ -209,271 +303,502 @@ impl fmt::Debug for PeriodicReader {
+            Err(e) => {
                 otel_error!(
-                    name: "collect_and_export",
-                    status = "timed_out"
+                    name: "PeriodReaderForceFlushError",
+                    event_name = "PeriodReaderForceFlushError",
+                    error = format!("{:?}", e)
                 );
-                Err(MetricsError::Other("export timed out".into()))
+                return Err(MetricsError::Other(e.to_string()));
             }
         }
-    }
 
-    async fn process_message(&mut self, message: Message) -> bool {
-        match message {
-            Message::Export => {
-                if let Err(err) = self.collect_and_export().await {
-                    global::handle_error(err)
-                }
+        if let Ok(response) = response_rx.recv() {
+            if response {
+                Ok(())
+            } else {
+                Err(MetricsError::Other("Failed to flush".into()))
             }
-            Message::Flush(ch) => {
-                let res = self.collect_and_export().await;
-                if ch.send(res).is_err() {
-                    global::handle_error(MetricsError::Other("flush channel closed".into()))
-                }
+        } else {
+            Err(MetricsError::Other("Failed to flush".into()))
+        }
+    }
+
+    fn shutdown(&self) -> Result<()> {
+        if self.is_shutdown.load(std::sync::atomic::Ordering::Relaxed) {
+            return Err(MetricsError::Other("reader is already shut down".into()));
+        }
+
+        let (response_tx, response_rx) = mpsc::channel();
+        match self.message_sender.lock() {
+            Ok(sender) => {
+                sender
+                    .send(Message::Shutdown(response_tx))
+                    .map_err(|e| MetricsError::Other(e.to_string()))?;
             }
-            Message::Shutdown(ch) => {
-                let res = self.collect_and_export().await;
-                let _ = self.reader.exporter.shutdown();
-                if ch.send(res).is_err() {
-                    global::handle_error(MetricsError::Other("shutdown channel closed".into()))
-                }
-                return false;
+            Err(e) => {
+                otel_error!(
+                    name: "PeriodReaderShutdownError",
+                    event_name = "PeriodReaderShutdownError",
+                    error = format!("{:?}", e)
+                );
+                return Err(MetricsError::Other(e.to_string()));
             }
         }
 
-        true
-    }
-
-    async fn run(mut self, mut messages: impl FusedStream<Item = Message> + Unpin) {
-        while let Some(message) = messages.next().await {
-            if !self.process_message(message).await {
-                break;
+        if let Ok(response) = response_rx.recv() {
+            self.is_shutdown
+                .store(true, std::sync::atomic::Ordering::Relaxed);
+            if response {
+                Ok(())
```

Comment:
Join the background thread before considering it a successful shutdown?

Reply:
Join the background thread could block the user thread if the background thread is some how blocking (like blocking in the export call), so the timeout parameter could be ignored.

Could we try to join it if it can be done within the specified timeout, and give up the joining operation once timeout is reached?

Reply:
@ThomsonTan Such a situation will not arise in current implementation. If `response_rx` got the message back, the background thread will be exiting itself, and not doing anything else.